### PR TITLE
server: refactor the independent service check

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: ref #8477.

### What is changed and how does it work?

This is a part of https://github.com/tikv/pd/pull/8517.

```commit-message
```

### Check List

Tests

- Unit test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -1417,6 +1417,18 @@ func (s *Server) GetRaftCluster() *cluster.RaftCluster {
...
 }
 
+// IsServiceIndependent returns whether the service is independent.
+func (s *Server) IsServiceIndependent(name string) bool {
+	if s.mode == APIServiceMode && !s.IsClosed() {
+		// TODO: remove it after we support tso discovery
+		if name == constant.TSOServiceName {
+			return true
+		}
+		return s.cluster.IsServiceIndependent(name)
```
Comment (lhy1024) [inline server/server.go:1427]: Do we need to check whether cluster is nil?

File: server/forward.go [hunk 2]
```diff
@@ -133,66 +133,82 @@ func (s *GrpcServer) forwardTSO(stream pdpb.PD_TsoServer) error {
 			err = errs.ErrGenerateTimestamp.FastGenByArgs("tso count should be positive")
 			return status.Error(codes.Unknown, err.Error())
 		}
-
-		forwardedHost, ok := s.GetServicePrimaryAddr(stream.Context(), constant.TSOServiceName)
-		if !ok || len(forwardedHost) == 0 {
-			tsoStreamErr = errors.WithStack(ErrNotFoundTSOAddr)
+		forwardCtx, cancelForward, forwardStream, lastForwardedHost, tsoStreamErr, err = s.handleTSOForwarding(forwardCtx, forwardStream, stream, server, request, tsDeadlineCh, lastForwardedHost, cancelForward)
```
Comment (lhy1024) [inline server/forward.go:136]: There are so many incoming and returning parameters here

**Review by lhy1024 — COMMENTED** [review]
the rest LGTM

File: server/forward.go (hunk 2 above)
Comment (rleungx) [reply server/forward.go:136]: Yes. But I don't find a better way to avoid it.

File: server/server.go (hunk 1 above)
Comment (rleungx) [reply server/server.go:1427]: no need

File: server/forward.go (hunk 2 above)
Comment (okJiang) [reply server/forward.go:136]: Is there any benefit after wraping `handleTSOForwarding`?

Comment (rleungx) [reply server/forward.go:136]: for upcoming PR

File: server/forward.go [hunk 3]
```diff
@@ -133,66 +133,82 @@ func (s *GrpcServer) forwardTSO(stream pdpb.PD_TsoServer) error {
...
-		}
+	response := &pdpb.TsoResponse{
+		Header: &pdpb.ResponseHeader{
+			ClusterId: tsopbResp.GetHeader().GetClusterId(),
+			Error:     pdpbErr,
+		},
+		Count:     tsopbResp.GetCount(),
+		Timestamp: tsopbResp.GetTimestamp(),
+	}
+	if server != nil {
```
[the remark this answers is not in the document]
Comment (rleungx) [reply server/forward.go:206]: We receive both `stream pdpb.PD_TsoServer` and `server *tsoServer` as the parameters.

[the remark this answers is not in the document]
Comment (okJiang) [reply server/forward.go:206]: `Server` seems that will not be `nil`?🤔

**Review by lhy1024 — APPROVED** [review]

**Review by okJiang — APPROVED** [review]

**Pull request merged**